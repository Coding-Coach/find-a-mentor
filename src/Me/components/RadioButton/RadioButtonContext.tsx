import { createContext } from 'react';

type RadioButtonContextProps =
  | undefined
  | {
      onChange(newValue: string): void;
      groupValue: string;
    };

export const RadioButtonContext = createContext<RadioButtonContextProps>(
  undefined
);
