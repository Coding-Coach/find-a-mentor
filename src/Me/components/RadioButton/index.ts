export { RadioButton as default } from './RadioButton';
export { default as RadioButtonGroup } from './RadioButtonGroup';
