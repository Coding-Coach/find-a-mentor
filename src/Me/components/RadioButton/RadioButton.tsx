import { InputHTMLAttributes, ReactChild, useContext } from 'react';
import styled from 'styled-components';
import { RadioButtonContext } from './RadioButtonContext';

type RadioButtonProps = {
  LabelComponent: ReactChild;
  checked?: boolean;
  onChange?: (value: string) => void;
};

export const RadioButtonContainer = styled.div`
  width: fit-content;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  cursor: pointer;
  position: relative;
  vertical-align: middle;

  & + & {
    margin-top: 10px;
  }
`;

const HiddenRadioButton = styled.input.attrs({ type: 'radio' })`
  border: 0;
  width: 24px;
  height: 24px;
  cursor: pointer;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
`;

const StyledRadioButton = styled.span<Pick<RadioButtonProps, 'checked'>>`
  border: 2px solid #69d5b1;
  border-radius: 50%;
  width: 24px;
  height: 24px;
  background-color: #fff;
  cursor: pointer;
  position: relative;
  display: block;

  input:checked + &:before {
    content: '';
    position: absolute;
    inset: 3px;
    background: #179a6f;
    border-radius: 50%;
    display: block;
  }
`;

const Label = styled.label`
  margin-left: 0.9rem;
  color: #4f4f4f;
  font-size: 14px;
  cursor: pointer;
  display: flex;
  align-items: center;
  gap: 5px;
`;

export const RadioButton = ({
  LabelComponent,
  checked,
  value,
  ...props
}: InputHTMLAttributes<HTMLInputElement> & RadioButtonProps) => {
  const groupContext = useContext(RadioButtonContext);
  const { groupValue, onChange } = groupContext || {
    onChange: props.onChange,
  };
  const defaultChecked = Boolean(groupContext ? groupValue === value : checked);

  return (
    <RadioButtonContainer>
      <Label>
        <HiddenRadioButton
          defaultChecked={defaultChecked}
          {...props}
          value={value}
          onChange={e => {
            onChange?.(e.target.value);
          }}
          type="radio"
        />
        <StyledRadioButton />
        {LabelComponent}
      </Label>
    </RadioButtonContainer>
  );
};
