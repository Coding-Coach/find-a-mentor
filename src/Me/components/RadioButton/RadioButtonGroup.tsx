import { FC, useState } from 'react';
import styled from 'styled-components';
import { RadioButtonContainer } from './RadioButton';
import { RadioButtonContext } from './RadioButtonContext';

const StyledRadioButtonGroup = styled.div`
  ${RadioButtonContainer} + ${RadioButtonContainer} {
    margin-top: 10px;
  }
`;

export type RadioButtonGroupProps = {
  value: string;
  onChange(value: string): void;
};

const RadioButtonGroup: FC<RadioButtonGroupProps> = ({
  value: defaultValue,
  onChange,
  children,
}) => {
  const [value, setValue] = useState(defaultValue);

  const onRadioButtonChange = (newValue: string) => {
    setValue(newValue);
    onChange(newValue);
  };

  return (
    <RadioButtonContext.Provider
      value={{ onChange: onRadioButtonChange, groupValue: value }}
    >
      <StyledRadioButtonGroup>{children}</StyledRadioButtonGroup>
    </RadioButtonContext.Provider>
  );
};

export default RadioButtonGroup;
