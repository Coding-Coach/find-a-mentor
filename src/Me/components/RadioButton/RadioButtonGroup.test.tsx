import { render } from '@testing-library/react';
import RadioButton from '.';
import RadioButtonGroup from './RadioButtonGroup';

describe('RadioButtonGroup', () => {
  it('should call onChange when check an option', async () => {
    const onChange = jest.fn();

    const { findByText } = render(
      <RadioButtonGroup value="2" onChange={onChange}>
        <RadioButton name="options" value="1" LabelComponent="Option 1" />
        <RadioButton name="options" value="2" LabelComponent="Option 2" />
        <RadioButton name="options" value="3" LabelComponent="Option 3" />
      </RadioButtonGroup>
    );
    (await findByText('Option 3')).click();

    expect(onChange).toHaveBeenCalledWith('3');
  });

  it('should match snapshot', () => {
    const { container } = render(
      <RadioButtonGroup value="2" onChange={() => {}}>
        <RadioButton name="options" value="1" LabelComponent="Option 1" />
        <RadioButton name="options" value="2" LabelComponent="Option 2" />
        <RadioButton name="options" value="3" LabelComponent="Option 3" />
      </RadioButtonGroup>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
