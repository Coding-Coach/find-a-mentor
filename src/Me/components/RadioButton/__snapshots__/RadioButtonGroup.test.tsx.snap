// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonGroup should match snapshot 1`] = `
.c2 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  vertical-align: middle;
}

.c1 + .c1 {
  margin-top: 10px;
}

.c4 {
  border: 0;
  width: 24px;
  height: 24px;
  cursor: pointer;
  opacity: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
}

.c6 {
  border: 2px solid #69d5b1;
  border-radius: 50%;
  width: 24px;
  height: 24px;
  background-color: #fff;
  cursor: pointer;
  position: relative;
  display: block;
}

input:checked + .c5:before {
  content: '';
  position: absolute;
  inset: 3px;
  background: #179a6f;
  border-radius: 50%;
  display: block;
}

.c3 {
  margin-left: 0.9rem;
  color: #4f4f4f;
  font-size: 14px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 5px;
}

.c0 .c1 + .c1 {
  margin-top: 10px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <label
      class="c3"
    >
      <input
        class="c4"
        name="options"
        type="radio"
        value="1"
      />
      <span
        class="c5 c6"
      />
      Option 1
    </label>
  </div>
  <div
    class="c1 c2"
  >
    <label
      class="c3"
    >
      <input
        checked=""
        class="c4"
        name="options"
        type="radio"
        value="2"
      />
      <span
        class="c5 c6"
      />
      Option 2
    </label>
  </div>
  <div
    class="c1 c2"
  >
    <label
      class="c3"
    >
      <input
        class="c4"
        name="options"
        type="radio"
        value="3"
      />
      <span
        class="c5 c6"
      />
      Option 3
    </label>
  </div>
</div>
`;
